File: ta_lib/strategies/signal/src/contrarian/kch_a.rs
use base::prelude::*;
use channel::a;
use core::prelude::*;
use timeseries::prelude::*;
use volatility::kch;

pub struct KchASignal {
    source: SourceType,
    smooth: Smooth,
    period: usize,
    smooth_atr: Smooth,
    period_atr: usize,
    factor: f32,
}

impl KchASignal {
    pub fn new(
        source: SourceType,
        smooth: Smooth,
        period: f32,
        smooth_atr: Smooth,
        period_atr: f32,
        factor: f32,
    ) -> Self {
        Self {
            source,
            smooth,
            period: period as usize,
            smooth_atr,
            period_atr: period_atr as usize,
            factor,
        }
    }
}

impl Signal for KchASignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.period_atr)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let source = data.source(self.source);
        let (upper, _, lower) = kch(
            &source,
            self.smooth,
            &data.atr(self.smooth_atr, self.period_atr),
            self.period,
            self.factor,
        );

        a!(upper, lower, source)
    }
}

File: strategy/generator/signal/contrarian/kch_a.py
from dataclasses import dataclass

from core.models.parameter import Parameter, RandomParameter, StaticParameter
from core.models.smooth import Smooth, SmoothATR
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class KchASignal(Signal):
    type: SignalType = SignalType.KchA
    source: Parameter = StaticParameter(SourceType.CLOSE)
    smooth: Parameter = StaticParameter(Smooth.UTLS)
    period: Parameter = RandomParameter(20.0, 60.0, 10.0)
    smooth_atr: Parameter = StaticParameter(SmoothATR.UTLS)
    period_atr: Parameter = RandomParameter(20.0, 80.0, 10.0)
    factor: Parameter = RandomParameter(0.3, 2.0, 0.1)
